text/0100-xcm-multi-type-asset-transfer.md

```diff
@@ -0,0 +1,301 @@
+///
+/// Assets in the holding register are matched using the given list of `AssetTransferFilter`s,
+/// they are then transferred based on their specified transfer type:
+///
+/// - teleport: burn local assets and append a `ReceiveTeleportedAsset` XCM instruction to
+///   the XCM program to be sent onward to the `dest` location,
+///
+/// - reserve deposit: place assets under the ownership of `dest` within this consensus system
+///   (i.e. its sovereign account), and append a `ReserveAssetDeposited` XCM instruction
+///   to the XCM program to be sent onward to the `dest` location,
+///
+/// - reserve withdraw: burn local assets and append a `WithdrawAsset` XCM instruction
+///   to the XCM program to be sent onward to the `dest` location,
+///
+/// The onward XCM is then appended a `ClearOrigin` to allow safe execution of any following
+/// custom XCM instructions provided in `remote_xcm`.
+///
+/// The onward XCM also potentially contains a `BuyExecution` instruction based on the presence
+/// of the `remote_fees` parameter (see below).
+///
+/// If a transfer requires going through multiple hops, an XCM program can compose this instruction
+/// to be used at every chain along the path, describing that specific leg of the transfer.
+///
+/// Parameters:
+/// - `dest`: The location of the transfer next hop.
+/// - `remote_fees`: If set to `Some(asset_xfer_filter)`, the single asset matching
+///   `asset_xfer_filter` in the holding register will be transferred first in the remote XCM
+///   program, followed by a `BuyExecution(fee)`, then rest of transfers follow.
+///   This guarantees `remote_xcm` will successfully pass a `AllowTopLevelPaidExecutionFrom` barrier.
+/// - `remote_xcm`: Custom instructions that will be executed on the `dest` chain. Note that
+///   these instructions will be executed after a `ClearOrigin` so their origin will be `None`.
+///
+/// Safety: No concerns.
+///
+/// Kind: *Command*.
+///
+InitiateAssetsTransfer {
+	destination: Location,
+	assets: Vec<AssetTransferFilter>,
+	remote_fees: Option<AssetTransferFilter>,
+	remote_xcm: Xcm<()>,
+}
+```
+
+An `InitiateAssetsTransfer { .. }` instruction shall transfer to `dest`, all assets in the `holding` register
+that match the provided `assets` and `remote_fees` filters.
+These filters identify the assets to be transferred as well as the transfer type to be used for transferring
+them.
+It shall handle the local side of the transfer, then forward an onward XCM to `dest` for handling
+the remote side of the transfer.
+
+It should do so using same mechanisms as existing `DepositReserveAsset`, `InitiateReserveWithdraw`, `InitiateTeleport`
+instructions but practically combining all required XCM instructions to be remotely executed into a _single_
+remote XCM program to be sent over to `dest`.
+
+Furthermore, through `remote_fees: Option<AssetTransferFilter>`, it shall allow specifying a single asset to be used
```

Comment:
Just an observation at the implementation level, when `dest` is a bridged location, for example, Ethereum; this instruction does initiate a multi-hop transfer. There is a hidden XCM fee charged by BridgeHub for committing `remote_xcm` for delivery to `dest`, and consequently the AH executor (SovereignPaidRemoteExporter) needs to estimate and take fees to cover that, with a degree of overestimation as xcm fees are variable).

Given that it likely that AH and BH will be collapsed into a single chain (Plaza) at some point in the future (thanks to CoreTime), I think its ok not to complicate the interface of `InitiateAssetsTransfer` with this complexity. 👍

Reply:
The last commit is exactly trying to clarify that the “dest” on the instruction is not the final destination, just the “next hop” from the point of view of assets movements (e.g. through a reserve chain). It is the destination for the current step of the transfer. Not the final destination.

----

> when dest is a bridged location like Ethereum; this instruction does end up initiating a multi-hop transfer that is somewhat hidden from the builder.

yes but it is “multi-hop” from purely a transport point of view.

from the actual XCM program point of view it is still “directly” AH->Ethereum. The XCM “origin” as seen by Ethereum is AH and the xcm program logic operates as if message was passed directly from AH.